### PR TITLE
feat(framework) Allow passing `torch.Tensor` to `Array` constructor

### PR DESCRIPTION
**Pull request by panh99**
Based on #4918

image

### CODE REVIEW

**Review by chongshenng — COMMENTED** [review]
LGTM!

File: src/py/flwr/common/record/parametersrecord_test.py [hunk 1]
```diff
@@ -38,9 +41,43 @@ def _get_buffer_from_ndarray(array: NDArray) -> bytes:
...
+
+
+class TfTensor(Mock):
+    """Mock TensorFlow tensor class."""
+
+
+MOCK_TORCH_TENSOR = TorchTensor(numpy=lambda: np.array([[1, 2, 3]]))
+MOCK_TF_TENSOR = TfTensor(numpy=lambda: np.array([1, 2, 3]))
+MOCK_TORCH_TENSOR.detach.return_value = MOCK_TORCH_TENSOR
+MOCK_TORCH_TENSOR.cpu.return_value = MOCK_TORCH_TENSOR
```
Comment (jafermarq) [inline src/py/flwr/common/record/parametersrecord_test.py:55]: Wouldn't it be better to introduce tf/torch into these tests instead of using mock? Although i don't think it's strictly needed for the tests here performed (just creating an `Array` obj) i think it's going to be needed once we introduce methods or functions to convert those `Arrays` back to tf/torch. 

It might be worth taking a look at the tests torch/tf themselves have when converting to NumPy:
- https://github.com/pytorch/pytorch/blob/main/test/test_numpy_interop.py
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/numpy_compat.py

File: src/py/flwr/common/record/parametersrecord.py [hunk 2]
```diff
@@ -171,6 +204,32 @@ def from_numpy_ndarray(cls, ndarray: NDArray) -> Array:
...
+        """Create Array from PyTorch tensor."""
+        if not (torch := sys.modules.get("torch")):
+            raise RuntimeError(
+                f"PyTorch is required to use {cls.from_torch_tensor.__name__}"
+            )
+
+        assert isinstance(
+            tensor, torch.Tensor
+        ), f"Expected PyTorch Tensor, got {type(tensor)}"
+        return cls.from_numpy_ndarray(tensor.detach().cpu().numpy())
```
[the remark this answers is not in the document]
Comment (panh99) [reply src/py/flwr/common/record/parametersrecord.py:218]: @jafermarq Just out of curiosity, may I know why we need to mention that? Is it b/c `detach()` or `cpu()` may cause some issues/overheads, or break anything?

File: src/py/flwr/common/record/parametersrecord_test.py (hunk 1 above)
Comment (panh99) [reply src/py/flwr/common/record/parametersrecord_test.py:55]: Yeah, it would be way easier to just use `torch` / `tensorflow` directly. However, we don't have `torch` or `tensorflow` in our dependencies. So I am not sure how we can directly use `torch` / `tensorflow` in unit tests.

Comment (jafermarq) [reply src/py/flwr/common/record/parametersrecord_test.py:55]: yes, after our discussion I agree better to have those as a e2e script 💯

File: src/py/flwr/common/record/parametersrecord.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jafermarq) [reply src/py/flwr/common/record/parametersrecord.py:218]: In principle nothing should break, specially if creating these records is something done just before the `ClientApp` is about to exit (e.g. returning the resulting message). But if someone is for example storing a tensor in the middle of training into the context, copying it to `.cpu()` can result in a slowdown (to be expected) if the tensor/model lives in CUDA. I think throwing a warning isn't practical. But if we can mention that the tensor is first "detached and moved to CPU" in the docstrings of this method then that would be more informative that having to look into the source code.

[the remark this answers is not in the document]
Comment (panh99) [reply src/py/flwr/common/record/parametersrecord.py:250]: That's a valid point. We should value transparency in this case. I've updated the PR, thanks!

**Review by jafermarq — DISMISSED** [review]

**Review by danieljanes — DISMISSED** [review]

**Review by danieljanes — APPROVED** [review]

**Review by danieljanes — DISMISSED** [review]

**Review by danieljanes — APPROVED** [review]

**Review by danieljanes — APPROVED** [review]
